### PR TITLE
machines: fix reading vCPU count from Libvirt's domain xml

### PR DESCRIPTION
**Pull request by mareklibra**

### CODE REVIEW

**Comment by mareklibra** [comment]
Failing tests are unrelated

**Comment by stefwalter** [comment]
Is this something that should be tested in either a unit test or integration test?

**Comment by mareklibra** [comment]
The issue is related to hot (un)plug CPUs. The virtual machine is created with max CPU count, but the "current" attribute in domain.xml shows actual allocation.

This aspect was missed prior this fix. In general, it's not straightforward to simulate it and it won't appear in the scenario the test VM is created.
In my opinion, it would not pay off to build so complex integration scenario with the recent level of cockpit:machines functionality.

I'm adding test for the CPU count, maybe there will be some "weird' setup in the future causing this to fail.

**Comment by stefwalter** [comment]
@mareklibra Would you be able to look into the test failure on ubuntu-1604 in check-machines. See above:

```
# testBasic (check_machines.TestMachines)
#
Ignoring Resource Error: Operation canceled http://10.111.120.130:9090/cockpit/$a14415f2d3b476dc98bc08c652c25dd8bc84d7fe/machines/d3.js
WARNING  The guest's network configuration does not support PXE
Wrote TestMachines-testBasic-FAIL.png
Journal extracted to TestMachines-testBasic-10.111.120.130-FAIL.log
not ok 49 testBasic (check_machines.TestMachines) duration: 99s
Traceback (most recent call last):
  File "./verify/check-machines", line 59, in testBasic
    b.wait_in_text("{0}-state".format(vm1_id), "running")
  File "/build/cockpit/test/common/testlib.py", line 242, in wait_in_text
    return self.wait_js_func('ph_in_text', selector, text)
  File "/build/cockpit/test/common/testlib.py", line 209, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/build/cockpit/test/common/testlib.py", line 722, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/build/cockpit/test/common/testlib.py", line 745, in _invoke
    raise Error(res['error'])
Error: timeout
```

**Comment by mareklibra** [comment]
@stefwalter , Might the test VM run out of disk space? I added more checks into last commit.

**Comment by stefwalter** [comment]
Lets try https://github.com/cockpit-project/cockpit/pull/5866

**Comment by mareklibra** [comment]
Root cause:
error: internal error: unable to execute QEMU command 'cont': Resetting the Virtual Machine is required

**Comment by stefwalter** [comment]
> Root cause: error: internal error: unable to execute QEMU command 'cont': Resetting the Virtual Machine is required

Back when working on the Cockpit integration tests I found QEMU pause/resume to be unreliable. Is this a normal part of our workflow or tests? Can we remove pausing during the testing?

**Comment by mareklibra** [comment]
In the failing tests, the VM is "paused" right after it's creation.
Google says, there might be multiple reasons for that like low disk space (seems not to be our case) or misconfiguration of the host (I don't know yet)

**Comment by mareklibra** [comment]
None of recently failing tests is related to check-machines.
None of the tests was running on cockpit-9 machine where the previous tests were failing.

This round, successful testing was namely on: cockpit-7, cockpit-8, cockpit-10, cockpit-11, verifymachine4, verifymachine5

Next steps:
- give it another try and see what happens
- are the tests failing repeatedly failing on certain machines like the cockpit-9?
- is there a race, did the dummy 30 secs waiting at the beginning help?

**Comment by mareklibra** [comment]
2nd round:
- successful machines: releasemachine1,

**Comment by mareklibra** [comment]
The "--cpu host-model-only" did not help, still failing on machines cockpit-9 and higher.
Adding `cat /proc/cpuinfo`.

Does anyone know if these cockpit-9 or above are different from the others?

**Comment by mareklibra** [comment]
According to cpuinfo, the cockpit-7 (passing) and cockpit-9 (failing) machines are the same. But virt-install is failing on cockpit-9.

I will continue looking into this issue.
But to move forward, is it fine to skip the test if virt-install fails to create the dummy testing VM?

**Comment by mareklibra** [comment]
Ok, both centos-7 on cockpit-9 or rhel-7 on cockpit-10 passed.

Test preparation phase (means the `virt-install` call) is unreliable, fails nondeterministically.
Failures are unrelated to particular test VM (like 'verify/fedora-25' or 'verify/debian-8') or host (like 'cockpit-7' or 'cockpit-9'). I don't even see correlation in matrix of those two.

Resume/Reset of a failed VM does not help.
Network is up (according to `virsh net-list`), disk space is available, hosts are interchangeable (`/proc/cpuino`), package versions (kernel, qemu) seem to be unrelated (fails older/newer, then it passes), libvirt log is mute.

Journal states for a failed attempt (not checked on a passing one):

```
Feb 09 14:10:12 unassigned-hostname systemd[1]: Started Virtual machine log manager.
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.8898] manager: (vnet0): new Tun device (/org/freedesktop/NetworkManager/Devices/5)
Feb 09 14:10:12 unassigned-hostname kernel: device vnet0 entered promiscuous mode
Feb 09 14:10:12 unassigned-hostname kernel: virbr0: port 1(vnet0) entered listening state
Feb 09 14:10:12 unassigned-hostname kernel: virbr0: port 1(vnet0) entered listening state
Feb 09 14:10:12 unassigned-hostname systemd-udevd[6448]: Could not generate persistent MAC address for vnet0: No such file or directory
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.9153] devices added (path: /sys/devices/virtual/net/vnet0, iface: vnet0)
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.9153] device added (path: /sys/devices/virtual/net/vnet0, iface: vnet0): no ifupdown configuration found.
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.9162] device (vnet0): state change: unmanaged -> unavailable (reason 'connection-assumed') [10 20 41]
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.9180] keyfile: add connection in-memory (f2a3bc99-3c9d-4546-bf14-5b18512e80c5,"vnet0")
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.9190] device (vnet0): state change: unavailable -> disconnected (reason 'connection-assumed') [20 30 41]
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.9234] device (vnet0): Activation: starting connection 'vnet0' (f2a3bc99-3c9d-4546-bf14-5b18512e80c5)
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.9249] device (vnet0): state change: disconnected -> prepare (reason 'none') [30 40 0]
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.9267] device (vnet0): state change: prepare -> config (reason 'none') [40 50 0]
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.9275] device (vnet0): state change: config -> ip-config (reason 'none') [50 70 0]
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.9294] device (virbr0): bridge port vnet0 was attached
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.9297] device (vnet0): Activation: connection 'vnet0' enslaved, continuing activation
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.9307] device (vnet0): state change: ip-config -> secondaries (reason 'none') [70 90 0]
Feb 09 14:10:12 unassigned-hostname NetworkManager[884]: <info>  [1486667412.9315] device (vnet0): state change: secondaries -> activated (reason 'none') [90 100 0]
Feb 09 14:10:13 unassigned-hostname NetworkManager[884]: <info>  [1486667413.0304] device (vnet0): Activation: successful, device activated.
Feb 09 14:10:13 unassigned-hostname dbus[850]: [system] Activating via systemd: service name='org.freedesktop.nm_dispatcher' unit='dbus-org.freedesktop.nm-dispatcher.service'
Feb 09 14:10:13 unassigned-hostname systemd[1]: Starting Network Manager Script Dispatcher Service...
Feb 09 14:10:13 unassigned-hostname dbus[850]: [system] Successfully activated service 'org.freedesktop.nm_dispatcher'
Feb 09 14:10:13 unassigned-hostname systemd[1]: Started Network Manager Script Dispatcher Service.
Feb 09 14:10:13 unassigned-hostname nm-dispatcher[6470]: req:1 'up' [vnet0]: new request (1 scripts)
Feb 09 14:10:13 unassigned-hostname nm-dispatcher[6470]: req:1 'up' [vnet0]: start running ordered scripts...
Feb 09 14:10:13 unassigned-hostname systemd[1]: Reloading OpenBSD Secure Shell server.
Feb 09 14:10:13 unassigned-hostname sshd[1243]: Received SIGHUP; restarting.
Feb 09 14:10:13 unassigned-hostname systemd[1]: Reloaded OpenBSD Secure Shell server.
Feb 09 14:10:13 unassigned-hostname sshd[1243]: Server listening on 0.0.0.0 port 22.
Feb 09 14:10:13 unassigned-hostname sshd[1243]: Server listening on :: port 22.
Feb 09 14:10:13 unassigned-hostname audit[6469]: AVC apparmor="STATUS" operation="profile_replace" profile="unconfined" name="libvirt-f76aa5fc-b48f-46de-ab32-bb3de9513fe5" pid=6469 comm="apparmor_parser"
Feb 09 14:10:13 unassigned-hostname audit[6469]: AVC apparmor="STATUS" operation="profile_replace" profile="unconfined" name="libvirt-f76aa5fc-b48f-46de-ab32-bb3de9513fe5//qemu_bridge_helper" pid=6469 comm="apparmor_parser"
Feb 09 14:10:13 unassigned-hostname kernel: audit: type=1400 audit(1486667413.288:14): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="libvirt-f76aa5fc-b48f-46de-ab32-bb3de9513fe5" pid=6469 comm="apparmor_parser"
Feb 09 14:10:13 unassigned-hostname kernel: audit: type=1400 audit(1486667413.288:15): apparmor="STATUS" operation="profile_replace" profile="unconfined" name="libvirt-f76aa5fc-b48f-46de-ab32-bb3de9513fe5//qemu_bridge_helper" pid=6469 comm="apparmor_parser"
Feb 09 14:10:13 unassigned-hostname virtlogd[6443]: libvirt version: 1.3.1, package: 1ubuntu10.6 (Christian Ehrhardt <christian.ehrhardt@canonical.com> Tue, 22 Nov 2016 09:39:18 +0100)
Feb 09 14:10:13 unassigned-hostname virtlogd[6443]: hostname: unassigned-hostname.unassigned-domain
Feb 09 14:10:13 unassigned-hostname virtlogd[6443]: End of file while reading data: Input/output error
Feb 09 14:10:13 unassigned-hostname libvirtd[1248]: libvirt version: 1.3.1, package: 1ubuntu10.6 (Christian Ehrhardt <christian.ehrhardt@canonical.com> Tue, 22 Nov 2016 09:39:18 +0100)
Feb 09 14:10:13 unassigned-hostname libvirtd[1248]: hostname: unassigned-hostname.unassigned-domain
Feb 09 14:10:13 unassigned-hostname libvirtd[1248]: internal error: unable to execute QEMU command 'cont': Resetting the Virtual Machine is required
Feb 09 14:10:14 unassigned-hostname libvirtd[6538]: libvirt version: 1.3.1, package: 1ubuntu10.6 (Christian Ehrhardt <christian.ehrhardt@canonical.com> Tue, 22 Nov 2016 09:39:18 +0100)
Feb 09 14:10:14 unassigned-hostname libvirtd[6538]: hostname: unassigned-hostname.unassigned-domain
Feb 09 14:10:14 unassigned-hostname libvirtd[6538]: Failed to open file '/sys/kernel/security/apparmor/profiles': Permission denied
Feb 09 14:10:14 unassigned-hostname libvirtd[6538]: Failed to read AppArmor profiles list '/sys/kernel/security/apparmor/profiles': Permission denied
Feb 09 14:10:14 unassigned-hostname kernel: virbr0: port 1(vnet0) entered learning state
Feb 09 14:10:16 unassigned-hostname ntpd[1378]: Listen normally on 7 vnet0 [fe80::fc54:ff:fe83:1387%6]:123
Feb 09 14:10:16 unassigned-hostname ntpd[1378]: new interface(s) found: waking up resolver
Feb 09 14:10:16 unassigned-hostname NetworkManager[884]: <info>  [1486667416.9017] device (virbr0): link connected
Feb 09 14:10:16 unassigned-hostname kernel: virbr0: topology change detected, propagating
Feb 09 14:10:16 unassigned-hostname kernel: virbr0: port 1(vnet0) entered forwarding state
Feb 09 14:10:18 unassigned-hostname ntpd[1378]: Listen normally on 8 virbr0 192.168.122.1:123
```

My suspicion is a race condition with some previous test changing VM configuration causing the domain failing to start.
The order of test calls (check-* scripts) is nondeterministic.

Our goal is to test the cockpit:machines, not the test environment itself or the virt-install.

With respect to that, I will change the test to:
- try to create a test VM (`virt-install`)
- check whether the VM is listed and a few its properties (the VM does not need to be running, just listed)
- if `virt-install` fails, check whether the state is 'paused' (conforms the defined virt-install behavior)
- otherwise check for the 'running' state and try to shut it down and check result

So in both cases, the test covers the main scenario, whether the application can be started, data retrieved, and React/Redux is working and data are refreshed according to system change (from "No VM state" to "A VM is registered").

**Comment by mareklibra** [comment]
Recently failing tests are unrelated.
The check-machines is implemented as stated in previous comment.

**Comment by mareklibra** [comment]
Can we agree on this scope, please?

**Comment by martinpitt** [comment]
In the debian-8 test, the `check-machines TestMachines.testBasic` test case still failed:
```
error: Failed to start network default
error: Unable to create bridge virbr0: File exists


 Name                 State      Autostart     Persistent
----------------------------------------------------------
 default              inactive   yes           yes

Ignoring Resource Error: Operation canceled http://10.111.126.151:9090/cockpit/$f74840670f78938443a3bfdcc8dc64ac5b40de37/machines/d3.js
```
Is this related to these changes? That said, the check-machines test already seems to be much more stable with this; it's currently almost impossible to get it to pass in other PRs (such as #5905). Thanks!

**Comment by mareklibra** [comment]
@martinpitt , These changes are trying to fix test environment setup and stability was improved.
Unfortunately, the `libvirtd` startup still fails nondeterministically (not related to particular host/test vm), repeated execution might pass.

What I propose: if test env setup fails, then the test should pass with appropriate log message.

Means, overall test execution will not be affected by false-negative results.
The cockpit:machines will be still tested since there will be still other test attempts with successful test env setup leading to actual test execution.

**Comment by martinpitt** [comment]
Yes, I didn't mean to reject these changes based on the failed test case -- improving stability is always welcome, doesn't need to become perfect in one step :-) I was just wondering if the more verbose test failure meant anything useful to you. I wouldn't go as far as marking the test as passed when the test env fails, that just hides errors and regressions.

**Comment by mareklibra** [comment]
Actually yes - it's useful. At least I know the test failure is caused by the env setup and it's cause - this time libvirt's networking is not working.
During other failed executions, the network is active, but actual VM creation fails. Unfortunately, I didn't find corresponding root cause for that.

I agree marking the test as passed when the test env fails was not a good idea.

I'll push other patch separating test env setup from the actual test, so it will be more clear in the code. The test will fail with failing setup phase.

**Comment by mareklibra** [comment]
So it seems to be working.
TL;DR: the test env setup is not stable and so far it might fail on:
- libvirtd network (the 'default' is inactive and attempt to restart fails anyway).
  In that case, the test fails with corresponding ERROR message
- virt-install can't properly create a testing nested vm
  In that case, the cockpit:machines are tested sufficiently - just the active operations are skipped from test (so far 'shut down VM')

The test env setup fails independently on host/test VM.
Our goal is to test the cockpit:machines, not the libvirt or nested-VM creation itself.
In most cases, the test environment setup passes ok so the plugin test is executed in it's full scope.

If the setup env setup can't be fully stabilized, we might either
- keep the behavior as described and implemented
- mock the `virsh` command, which I do not prefer since behavior of `virsh` might change in the course of time.

**Comment by martinpitt** [comment]
7/10 and none of the failures are in TestMachines* any more, nice! I restarted the three failing ones, to collect some more data. Nice work, thanks!

File: pkg/machines/hostvmslist.jsx [hunk 1]
```diff
@@ -204,10 +204,10 @@ const VmOverviewTab = ({ vm }) => {
             <tr className='machines-listing-ct-body-detail'>
                 <td>
                     <table className='form-table-ct'>
-                        <VmOverviewTabRecord id={`${vmId(vm.name)}-state`} descr='State:' value={vm.state}/>
+                        <VmOverviewTabRecord id={`${vmId(vm.name)}-state`} descr={_("State:")} value={vm.state}/>
```
Comment (martinpitt) [inline pkg/machines/hostvmslist.jsx]: This looks like "State" would get translated, is that on purpose?

File: pkg/machines/libvirt.es6 [hunk 2]
```diff
@@ -226,10 +226,16 @@ function spawnVirshReadOnly({connectionName, method, name, failHandler}) {
 function parseDumpxml(dispatch, connectionName, domXml) {
     const xmlDoc = $.parseXML(domXml);
 
+    if (!xmlDoc) {
+        console.error(`Can't parse dumpxml, input: "${domXml}"`);
+        return ;
+    }
```
Comment (martinpitt) [inline pkg/machines/libvirt.es6]: This looks like it would silently ignore parsing errors. I don't claim to understand the details of what's going on here, but this looks a bit dubious? @stefwalter , are you familiar with this part?

File: pkg/machines/hostvmslist.jsx (hunk 1 above)
Comment (mareklibra) [inline pkg/machines/hostvmslist.jsx]: Yes, it's a table column header string

File: pkg/machines/libvirt.es6 (hunk 2 above)
Comment (mareklibra) [inline pkg/machines/libvirt.es6]: This really should not happen :-) 
It means, the `virsh dumpxml` produced unparsable output or there was error in transfer - I noticed transfer of an empty string.
With this "handling", subsequent data retrieval is executed so there's still a chance to get reasonable data and during next polling interval the `dumpxml` is reexecuted.

Other option is to fail as before this change, so an exception is thrown and the cockpit's error messages in the right top corner are populated.
I would not introduce other error messages printed on the page.

**Comment by dperpeet** [comment]
I did some work on the machines tests on top of this, will post follow-up

**Comment by dperpeet** [comment]
replaced by #5916
